src/python/pants/backend/go/target_type_rules.py

```diff
@@ -0,0 +1,47 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from pants.backend.go.target_types import GoModuleSources, GoPackageDependencies
+from pants.base.specs import AddressSpecs, AscendantAddresses, SiblingAddresses
+from pants.build_graph.address import Address
+from pants.engine.internals.selectors import Get
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import (
+    InjectDependenciesRequest,
+    InjectedDependencies,
+    UnexpandedTargets,
+    WrappedTarget,
+)
+from pants.engine.unions import UnionRule
+
+
+class InjectGoModuleDependency(InjectDependenciesRequest):
+    inject_for = GoPackageDependencies
+
+
+@rule
+async def inject_go_module_dependency(request: InjectGoModuleDependency) -> InjectedDependencies:
+    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
+    spec_path = original_tgt.target.address.spec_path
+    candidate_targets = await Get(
+        UnexpandedTargets,
+        AddressSpecs([AscendantAddresses(spec_path), SiblingAddresses(spec_path)]),
+    )
+    go_module_targets = [
+        (tgt.address.spec_path, tgt)
+        for tgt in candidate_targets
+        if tgt.has_field(GoModuleSources) and not tgt.address.is_file_target
+    ]
+    sorted_go_module_targets = sorted(go_module_targets, key=lambda x: x[0], reverse=True)
```

Comment:
Simpler is probably to not store a tuple in the above collection, and sort by `lambda tgt: tgt.address.spec_path`. That should simplify a couple lines in this rule